### ISSUE TITLE
About the process of training a new model

### ISSUE THREAD

**Issue by daobilige-su** [meijieru/crnn.pytorch#2]
Hi meijieru,

First of all, thanks so much for the package, it works nicely and is easy to install and test.

I am really interested in your package and I am trying to go deeper by redoing the whole process starting from training the model. I am quite a beginner in this field, so I have some question as follows,

1. If I understand correctly, to train a model, I need to prepare a training dataset (a set of images of words and their corresponding labels) and pass it to "create_dataset.py" of the original "crnn" package. Then I have to use "convert_dataset.py" file in the "tool" folder to sort the training files according to their width. Finally use "crnn_main.py" to train the model. Is that correct?

2. Is possible, is there anywhere that I can download the training data to reproduce the exactly same (or similar) result (final trained model) as you got in the end? 

I knew these questions might sound silly to you, but your comments are greatly appreciated. I think it is also helpful to others who are new to this field. Many thanks.

Su

**Comment by meijieru** [comment]
1. You can easily modified the `create_dataset.py` to sort the image when constructed. I do this pipeline because the dataset is too large to re-constructed, so I choose to convert it to meet my need.

2. This dataset is too large (around to 30G) to upload. I am really sorry for that.

**Comment by daobilige-su** [comment]
Thank you!

**Comment by bachhovuong310** [comment]
Hi meijieru,
Thank you for sharing the project. Now, I am trying to train model with another dataset like  _daobilige-suso_ but I have some different question:
1. How you train model with dataset have many types of length of words ? After training, my model have low accuracy with very short word or very long word. 
2. The model can train image have multi words ( ex: full name, location) ? I hope this model can learn white space between two words :)

Thank you very much

**Comment by meijieru** [comment]
1. You could first train with fixed-size images, and then fine tune with images which keep its origin ratio, by setting the corresponding flag.
2. It seems not able to deal with that issue, because CTC separate characters by blank. This model assumes the inputs are words without blanks.

**Comment by bachhovuong310** [comment]
Thank you Meijieru !!!

**Comment by ahmedmazari** [comment]
Hi @meijieru , let me first thank you for sharing your package.

l have some questions related to the ones asked by @bachhovuong310 :  
1.  l don't understand the second part answer -1-  **"and then fine tune with images which keep its origin ratio, by setting the corresponding flag."** 
Do you mean by that , we first of all need to train the model with fixed size images :  let's say images of 40 by 73  height= 40 and width= 73. Once we finished the training on the fixed size images. We feed the architecture  (to fine tune it) and train it again with variable size images such as images of (45,70), (37,56), (38,70), (96, 17), (98,30), is it that ?  

2- l have a black and white sequence images, pixels are either 0 or 255, but of variable height and width. Can this model be feeded with black and white images (where pixels are only 0 or 255) where the height and with of images are variable ? or the model can take these raw images as they are and pre-process  them to adapt them to the architecture requirement  ?

3-  in `demo.py` you set `alphabet = '0123456789abcdefghijklmnopqrstuvwxyz'` . I'm wondering if can change the alphabet to adapt to my data. l have sequences as follow :  
 3-A- Sequences of  transaction amount :   **12 . 34567 , 47**  `124,456` 
so l need to add to the alphabet  `comma` and `dot `
 3-B  Sequences of words  containing special characters :   ` , ; : ! / \ % $ à é è  à @ û ü .`
 3-C  Make  between Uppercase and lowercase  a A b B
so the alphabet will look like this : 
alphabet='0123456789abcdefghijklmnopkrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ,;:!/\%$àéèà@ûü'.'
Can the architecture support this alphabet ? 

Thank you for you responses, any comment would be appreciated

**Comment by meijieru** [comment]
1. If your image's size is 256x100, then scale it to 82(floor(256 / 100 * 32))x32
2. It should work.
3. That's OK.

**Comment by ahmedmazari** [comment]
@meijieru , thank you for your quick response.

1-
A- "If your image's size is 256x100, then scale it to 82(floor(256 / 100 * 32))x32" =>  l understand that the images should be of the height which is multiple of 32 , isn't it ?
B-  If yes, is that required just for` demo.py`  or even for training the model with our own dataset ?
C- When you rescale the images as you suggested, we will not encounter any information loss ? which decrease the accuracy of the model since the images are deformed ?

2- 

A- "then fine tune with images which keep its origin ratio, by setting the corresponding flag."
You mean, at the beginning, we should (have) train the model with a set of images which all of them have the same dimension x*y , then when we want to add other images we have to rescale them to the dimension x*y, isn't it ? 
B- If we don't have a sample of sequence images which has the same dimension. So we need to construct that by applying padding in order to get all the images to the same dimension. l can take for instance the max width and max height of all my images then apply padding to all the images to construct sequences images of dimension (max_height, max_width)  so there will be no problem of dimension. sequences are kept in the center of image and padding added to fit that dimension.  What do you think about it,  the  architecture  will learn from that kind of pre-processed data ?

3- Have you tried to feed your architecture with noisy sequences ( blur ,  characters hided a little bit behind a background color , noise ...) and it works  (if indeed we have the right lables )? 
 
Thank you a lot for you clarifications

**Comment by bachhovuong310** [comment]
@ahmedmazari 
2 - B I tryed the same pre-processed data method but  the result is not good as you scale the image to the fixed size.  
3 - I follow https://github.com/ankush-me/SynthText to make train data . It should work.

**Comment by ahmedmazari** [comment]
@bachhovuong310 ,
2-B so how have you solved the problem ?
3- what are the performance, after your trials , have you noticed that the architecture was able to  make good prediciton of noisy characters and blurred images ?

Thank you

**Comment by ahmedmazari-dhatim** [comment]
@daobilige-su , what are the outputs of create_dataset.py ? 
are they :  `data.mdb` and `lock.mdb` ?
If yes , are they the inputs for convert_dataset.py ?
l don't understand the link between cretae_dataaset.py and convert_datataset.py

Thank you
